L1Trigger/L1TTrackMatch/interface/L1TkHTMissEmulatorProducer.h

```diff
@@ -0,0 +1,219 @@
+
+  const unsigned int kPhiBins = 1 << kInternalPhiWidth;
+
+  const float kMaxCosLUTPhi{M_PI};
+
+  const string kLUTdir{"LUTs/"};
+
+
+  template <typename T>
+    T digitizeSignedValue(double value, unsigned int nBits, double lsb) {
+    T digitized_value = std::floor(std::abs(value) / lsb);
+    T digitized_maximum = (1 << (nBits - 1)) - 1;  // The remove 1 bit from nBits to account for the sign
+    if (digitized_value > digitized_maximum)
+      digitized_value = digitized_maximum;
+    if (value < 0)
+      digitized_value = (1 << nBits) - digitized_value;  // two's complement encoding
+    return digitized_value;
+  }
+
+
+  //Ouput LUTs to file for use in FW
+  template <typename T>
+    void writeLUTtoFile(vector<T>(&LUT), const string& filename, const string& delimiter) {
+    int fail = system((string("mkdir -p ") + l1tmhtemu::kLUTdir).c_str());
+    if (fail)
+      throw cms::Exception("BadDir") << __FILE__ << " " << __LINE__ << " could not create directory "
+                                     << l1tmhtemu::kLUTdir;
+
+    const string fname = l1tmhtemu::kLUTdir + filename + ".tab";
+    ofstream outstream(fname);
+    if (outstream.fail())
+      throw cms::Exception("BadFile") << __FILE__ << " " << __LINE__ << " could not create file " << fname;
+
+    outstream << "{" << endl;
+    for (unsigned int i = 0; i < LUT.size(); i++) {
+      if (i != 0)
+        outstream << delimiter << endl;
+      outstream << LUT[i];
+    }
+    outstream << endl << "};" << endl;
+    outstream.close();
+  }
+
+
+  std::vector<phi_t> generateCosLUT(unsigned int size) {  // Fill cosine LUT with integer values
+    float phi = 0;
+    std::vector<phi_t> cosLUT;
+    for (unsigned int LUT_idx = 0; LUT_idx < size; LUT_idx++) {
+      cosLUT.push_back(digitizeSignedValue<phi_t>(cos(phi), l1tmhtemu::kInternalPhiWidth, l1tmhtemu::kStepPhi));
+
+      phi += l1tmhtemu::kStepPhi;
+    }
+    cosLUT.push_back((phi_t)(0));  //Prevent overflow in last bin
+
+    return cosLUT;
+  }
+
+
+  std::vector<MHTphi_t> generateaTanLUT(int cordicSteps) {  // Fill atan LUT with integer values
+    std::vector<MHTphi_t> atanLUT;
+    for (int cordicStep = 0; cordicStep < cordicSteps; cordicStep++) {
+      atanLUT.push_back(MHTphi_t(round((kMHTPhiBins * atan(pow(2,-1*cordicStep))) / (2 * M_PI))));
+    }
+    return atanLUT;
+  }
+
+
+  std::vector<Et_t> generatemagNormalisationLUT(int cordicSteps) {
+    float val = 1.0;
+    std::vector<Et_t> magNormalisationLUT;
+    for (int cordicStep = 0; cordicStep < cordicSteps; cordicStep++) {
+      val = val / (pow(1 + pow(4, -1*cordicStep), 0.5));
+      magNormalisationLUT.push_back(Et_t(round(kMHTBins * val)));
+    }
+    return magNormalisationLUT;
+  }
+
+
+  struct EtMiss {
+    MHT_t Et;
+    MHTphi_t Phi;
+  };
+
+
+  EtMiss cordicSqrt(Et_t x, Et_t y, int cordicSteps) {
+
+    std::vector<l1tmhtemu::MHTphi_t> atanLUT;
+    atanLUT = generateaTanLUT(cordicSteps);
```

Comment:
Why not call `generateaTanLUT` and `generatemagNormalisationLUT` in the constructor? That way you don't have to regenerate them every event when `cordicSqrt` is called. Instead you could pass them by reference if you want to keep `cordicSqrt` outside of the class.

Reply:
`generateaTanLUT` and `generatemagNormalisationLUT` are now in constructor in the latest commit